torchtext/experimental/datasets/translation.py

```diff
@@ -21,72 +21,61 @@ def _setup_datasets(dataset_name,
-    else:
-        if not isinstance(src_vocab, Vocab):
-            raise TypeError("Passed src vocabulary is not of type Vocab")
-    logging.info('src Vocab has {} entries'.format(len(src_vocab)))
-
-    if tgt_vocab is None:
-        if 'train' not in data_select:
-            raise TypeError("Must pass a vocab if train is not selected.")
-        logging.info('Building tgt Vocab based on train data')
-        tgt_vocab = build_vocab(raw_data["train"],
-                                tgt_text_vocab_transform,
-                                index=1)
-    else:
-        if not isinstance(tgt_vocab, Vocab):
-            raise TypeError("Passed tgt vocabulary is not of type Vocab")
-    logging.info('tgt Vocab has {} entries'.format(len(tgt_vocab)))
+    if 'train' not in data_select and None in vocab:
+        raise TypeError("If train is not selected must pass Vocab for both source and target.")
+    if vocab[0] is not None and not isinstance(vocab[0], Vocab):
+        raise TypeError("Passed src vocabulary is of type %s expected type Vocab".format(type(src_vocab)))
+    if vocab[1] is not None and not isinstance(vocab[1], Vocab):
+        raise TypeError("Passed tgt vocabulary is of type %s expected type Vocab".format(type(tgt_vocab)))
+    if not isinstance(tokenizer, tuple) or len(tokenizer) != 2:
+        raise ValueError("tokenizer must be tuple of length two. One for "
+                         "source and target respectively. %s passed instead".format(tokenizer))
+    tokenizer = (
+        get_tokenizer("spacy", language='de_core_news_sm') if tokenizer[0] is None else tokenizer[0],
+        get_tokenizer("spacy", language='en_core_web_sm') if tokenizer[1] is None else tokenizer[1])
+
+    def build_raw_iter(raw_iter=None):
+        raw_iter_ = raw.translation.DATASETS[dataset_name](train_filenames=train_filenames,
+                                                           valid_filenames=valid_filenames,
+                                                           test_filenames=test_filenames,
+                                                           root=root, data_select=data_select)
+        if raw_iter is None:
+            raw_iter = {}
+        for i, name in enumerate(data_select):
+            if name not in raw_iter:
+                raw_iter[name] = raw_iter_[i]
+        return raw_iter
+    raw_iter = build_raw_iter()
+
+    # pop train iterator to force repopulation
+    vocab_ = len(vocab) * [None]
+    for i in range(len(vocab)):
+        vocab_[i] = build_vocab(raw_iter.pop("train"),
+                                tokenizer[i],
+                                index=i)
+        raw_iter = build_raw_iter()
+    vocab = tuple(vocab_)
+    logging.info('src Vocab has {} entries'.format(len(vocab[0])))
+    logging.info('tgt Vocab has {} entries'.format(len(vocab[1])))
+
+    raw_data = {}
+    for name in raw_iter:
+        raw_data[name] = list(raw_iter[name])
+
+    def build_transform(vocab, tokenizer):
```

Comment:
Is this same as `sequential_transforms`?

Reply:
No, because it allows tokenizer to be a generator.
```
def sequential_transforms(*transforms):
    def func(txt_input):
        for transform in transforms:
            txt_input = transform(txt_input)
        return txt_input
```

requires tokenizer to materialize the entire list of tokens, whereas here we're iterating over the token the tokenizer returns and are then looking it up immediately within the vocab without ever storing it in a list.